cdap-app-templates/cdap-program-report/cdap-program-report-base/src/test/java/co/cask/cdap/report/ReportGenerationAppTest.java

```diff
@@ -242,29 +214,143 @@ public boolean accept(String className, URL classUrl, URL classPathUrl) {
+    // from spark-avro so that spark-avro and its dependencies will be included.
+    bundler.createBundle(avroSparkBundle, DefaultSource.class);
+    File unJarDir = BundleJarUtil.unJar(avroSparkBundle, TEMP_FOLDER.newFolder());
+
+    ApplicationManager app = deployApplication(deployNamespace,
+                                               ReportGenerationApp.class, new File(unJarDir, "lib").listFiles());
+    return app.getSparkManager(ReportGenerationSpark.class.getSimpleName()).start(runtimeArguments);
+  }
+
+  @Test
+  public void testReportExpiration() throws Exception {
+    NamespaceId testNamespace = new NamespaceId("reporting");
+    NamespaceMeta reportingNamespace = new NamespaceMeta.Builder()
+      .setName(testNamespace)
+      .setDescription("Reporting namespace used to test repoort expiry")
+      .build();
+    getNamespaceAdmin().create(reportingNamespace);
+    long currentTimeMillis = System.currentTimeMillis();
+    DatasetId datasetId = createAndInitializeDataset(testNamespace, currentTimeMillis);
+    Map<String, String> runTimeArguments = new HashMap<>();
+    // set report expiration time to be 10 seconds
+    runTimeArguments.put(Constants.Report.REPORT_EXPIRY_TIME, "10");
+    SparkManager sparkManager = deployAndStartReportingApplication(testNamespace, runTimeArguments);
+
+    URL url = sparkManager.getServiceURL(1, TimeUnit.MINUTES);
+    Assert.assertNotNull(url);
+    URL reportURL = url.toURI().resolve("reports/").toURL();
+    List<Filter> filters =
+      ImmutableList.of(
+        new ValueFilter<>(Constants.NAMESPACE, ImmutableSet.of("ns1", "ns2"), null),
+        new RangeFilter<>(Constants.DURATION, new RangeFilter.Range<>(null, 500L)));
+    long startSecs = TimeUnit.MILLISECONDS.toSeconds(currentTimeMillis);
+    ReportGenerationRequest request =
+      new ReportGenerationRequest("ns1_ns2_report", startSecs, startSecs + 30,
+                                  new ArrayList<>(ReportField.FIELD_NAME_MAP.keySet()),
+                                  ImmutableList.of(new Sort(Constants.DURATION, Sort.Order.DESCENDING)), filters);
+    HttpURLConnection urlConn = (HttpURLConnection) reportURL.openConnection();
+    urlConn.setDoOutput(true);
+    urlConn.setRequestMethod("POST");
+    urlConn.getOutputStream().write(GSON.toJson(request).getBytes(StandardCharsets.UTF_8));
+    if (urlConn.getErrorStream() != null) {
+      Assert.fail(Bytes.toString(ByteStreams.toByteArray(urlConn.getErrorStream())));
+    }
+    Assert.assertEquals(200, urlConn.getResponseCode());
+
+    Map<String, String> reportIdMap = getResponseObject(urlConn, STRING_STRING_MAP);
+    String reportId = reportIdMap.get("id");
+    Assert.assertNotNull(reportId);
+    URL reportIdURL = reportURL.toURI().resolve("info?report-id=" + reportId).toURL();
+    validateReportSummary(reportIdURL, startSecs);
+
+    // perform get reports list for 60 seconds and make sure we get 0 reports due to expiry
+    Tasks.waitFor(0, () -> {
```

Comment:
since we have set expiry time to be 20 seconds, we want to retry and at the beginning the list reports will return size 1, while after it would return size 0 after expiration.